### PR TITLE
Adding `__rshift__` to `CircuitComponent`s

### PR DESCRIPTION
**Pull request by SamFerracin**
**Context:**
Adds ``>>`` to circuit components. A few other minimal changes to make it work

### CODE REVIEW

File: mrmustard/lab_dev/circuit_components.py [hunk 1]
```diff
@@ -233,6 +233,37 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
 
         return CircuitComponent.from_attributes("", wires_ret, representation_ret)
 
+    def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+        """
+        ret = self @ other
```
Comment (sylviemonet) [inline mrmustard/lab_dev/circuit_components.py:241]: make it more readable by removing the ret pls. 
raise an error in these two cases: self.bra with other.ket and vice versa.

Comment (SamFerracin) [reply mrmustard/lab_dev/circuit_components.py]: Thanks. Done, let me know if it's still hard to parse

File: mrmustard/lab_dev/circuit_components.py [hunk 2]
```diff
@@ -233,6 +233,42 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
...
+            # self has ket and bra, other has ket and bra
+            return self @ other
+
+        if self.wires.ket and self.wires.bra and other.wires.ket and not other.wires.bra:
+            # self has ket and bra, other has ket
+            return self @ other @ other.adjoint
+
+        if self.wires.ket and self.wires.bra and not other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has and bra
+            return self @ other @ other.adjoint
```
Comment (sylviemonet) [inline mrmustard/lab_dev/circuit_components.py:251]: would both of them can be merged?

 ``if self.wires.ket and self.wires.bra and (other.wires.ket or other.wires.bra):``

File: mrmustard/lab_dev/circuit_components.py [hunk 3]
```diff
@@ -233,6 +233,42 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
...
+
+        if self.wires.ket and not self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket, other has ket and bra
+            return self @ self.adjoint @ other
+
+        if not self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has bra, other has ket and bra
+            return self @ self.adjoint @ other
```
Comment (sylviemonet) [inline mrmustard/lab_dev/circuit_components.py:259]: The same, maybe it can merge with 

`if (self.wires.ket or self.wires.bra) and other.wires.ket and other.wires.bra:`

**Review by sylviemonet — COMMENTED** [review]
Nice! 
I have no problems with this PR except for small suggestions about the shift logic.

File: mrmustard/lab_dev/circuit_components.py [hunk 4]
```diff
@@ -233,6 +233,34 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
...
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+        """
+        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has ket and bra
+            return self @ other
+
+        if self.wires.ket and self.wires.bra and (other.wires.ket or other.wires.bra):
+            # self has ket and bra, other has ket or bra
+            return self @ other @ other.adjoint
```
Comment (ziofil) [inline mrmustard/lab_dev/circuit_components.py:247]: Does this work in either case? I mean either if `other` has only ket or only bra?

Comment (sylviemonet) [reply mrmustard/lab_dev/circuit_components.py:247]: This is something anti-heuristic to me. I don't know how to see an object like this. 

As I understand here, this is to add the adjoint according to if one has only bra to ket but the other one has both.

File: mrmustard/lab_dev/circuit_components.py [hunk 5]
```diff
@@ -233,6 +233,34 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
 
         return CircuitComponent.from_attributes("", wires_ret, representation_ret)
 
+    def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+        """
+        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has ket and bra
```
Comment (ziofil) [inline mrmustard/lab_dev/circuit_components.py:242]: i think the if statement is clear enough

File: mrmustard/lab_dev/circuit_components.py [hunk 6]
```diff
@@ -233,6 +233,34 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
...
+            return self @ other
+
+        if self.wires.ket and self.wires.bra and (other.wires.ket or other.wires.bra):
+            # self has ket and bra, other has ket or bra
+            return self @ other @ other.adjoint
+
+        if (self.wires.ket or self.wires.bra) and other.wires.ket and other.wires.bra:
```
Comment (ziofil) [inline mrmustard/lab_dev/circuit_components.py:249]: now that i think of it: isn't `self.wires.ket or self.wires.bra` always `True`?

File: mrmustard/lab_dev/states/base.py [hunk 7]
```diff
@@ -48,9 +46,28 @@ class DM(State):
...
+        r"""
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+
+        Returns a ``DM`` when ``other`` is a ``Unitary`` or a ``Channel``, and ``other`` acts on
+        ``self``'s modes. Otherwise, it returns a ``CircuitComponent``.
+        """
+        component = super().__rshift__(other)
+
+        if isinstance(other, (Unitary, Channel)) and set(other.modes).issubset(set(self.modes)):
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:64]:
```suggestion
        if isinstance(other, (Unitary, Channel)) and set(other.modes).issubset(self.modes):
```

File: mrmustard/lab_dev/circuit_components.py (hunk 6 above)
Comment (sylviemonet) [reply mrmustard/lab_dev/circuit_components.py:249]: Oh I think you are right. It would be one of them is True but not both of them...Emmmm needs a better condition here

File: tests/test_lab_dev/test_circuit_components.py [hunk 8]
```diff
@@ -169,6 +184,77 @@ def test_matmul_is_associative(self):
...
+                0.08944272 - 0.08944272j,
+                0.083666 - 0.083666j,
+                0.08944272 + 0.08944272j,
+                0.08944272 + 0.08944272j,
+                0.083666 + 0.083666j,
+            ],
+        )
+        assert np.allclose(result.representation.c, 0.95504196)
+
+    def test_rshift_is_associative(self):
```
Comment (ziofil) [inline tests/test_lab_dev/test_circuit_components.py:237]: love it

File: mrmustard/lab_dev/circuit_components.py (hunk 4 above)
Comment (SamFerracin) [reply mrmustard/lab_dev/circuit_components.py:247]: If other has ket AND bra, then line 243 returns. The return in line 247 happens if other has either ket or bra.
If you guys think that the version below is more readable, I can change it

```

        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
            # self has ket and bra, other has ket and bra
            return self @ other

        if self.wires.ket and self.wires.bra and other.wires.ket and not other.wires.bra:
            # self has ket and bra, other has ket
            return self @ other @ other.adjoint
```

**Review by ziofil — COMMENTED** [review]
I love how readable the tests are! A testament to a well made library :)
I left some questions.

File: mrmustard/lab_dev/circuit_components.py (hunk 4 above)
Comment (SamFerracin) [reply mrmustard/lab_dev/circuit_components.py:247]: @yuan I am not sure what you mean by anti-heuristic, but line 247 is what is returned when you do ``attenuator >> dgate``

File: mrmustard/lab_dev/circuit_components.py (hunk 6 above)
Comment (SamFerracin) [reply mrmustard/lab_dev/circuit_components.py:249]: We could, but that's already taken care of by the matmul, when it does ``self.wires @ other.wires`` to find the wires of the returned object. If the two objects have invalid input/output wires (for example, ``vacuum >> vacuum``), it will error

In other words, the error belongs to the lower level implementation

File: mrmustard/lab_dev/circuit_components.py (hunk 4 above)
Comment (sylviemonet) [reply mrmustard/lab_dev/circuit_components.py:247]: Anti-heuristic to me means one thing only has bra no ket..

Comment (ziofil) [reply mrmustard/lab_dev/circuit_components.py:246]: @sylviemonet That is true, but I think we should always implement the most general case.

Comment (ziofil) [reply mrmustard/lab_dev/circuit_components.py:246]: This is checking for bra/ket globally, should't we only check self.output other.input?

**Review by ziofil — APPROVED** [review]
super!

**Review by sylviemonet — APPROVED** [review]
